File: src/CVEzD3FEND/models/bundle.py
"""Bundle-level models per contracts/BUNDLE_CONTRACT.md and VALIDATION_CONTRACT.md."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from CVEzD3FEND.models.graph import Edge, Node

CoverageStatus = Literal["covered", "partial", "gap", "unknown", "not_applicable"]
SourceStatus = Literal["ok", "fallback", "unavailable", "error"]


class Source(BaseModel):
    source_id: str
    name: str
    kind: str
    url: str | None = None
    fetched_at: str
    version: str | None = None
    sha256: str | None = None
    record_count: int | None = None
    status: SourceStatus = "ok"
    compressed: bool = False
    license: str = "See docs/ATTRIBUTION.md"
    notes: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class Route(BaseModel):
    route_id: str
    start_node: str
    end_node: str
    path: list[str]
    nodes: list[str]
    edges: list[str]
    confidence: float
    canonical: bool
    inferred: bool
    coverage_status: CoverageStatus = "unknown"
    recommended_actions: list[str] = Field(default_factory=list)
    evidence_required: list[str] = Field(default_factory=list)
    source_refs: list[str] = Field(default_factory=list)


class CoverageTechnique(BaseModel):
    attack_technique: str
    defend_techniques: list[str] = Field(default_factory=list)
    controls: list[str] = Field(default_factory=list)
    detections: list[str] = Field(default_factory=list)
    evidence: list[str] = Field(default_factory=list)
    data_sources: list[str] = Field(default_factory=list)
    log_sources: list[str] = Field(default_factory=list)
    coverage_status: CoverageStatus = "unknown"
    gap_reason: str | None = None
    owner: str | None = None
    last_validated_at: str | None = None
    confidence: float = 1.0


class CoverageSummary(BaseModel):
    covered: int = 0
    partial: int = 0
    gap: int = 0
    unknown: int = 0
    not_applicable: int = 0


class Coverage(BaseModel):
    techniques: list[CoverageTechnique] = Field(default_factory=list)
    summary: CoverageSummary = Field(default_factory=CoverageSummary)


class Warning(BaseModel):
    code: str
    message: str
    context: dict[str, Any] = Field(default_factory=dict)


class QualityReport(BaseModel):
    generated_at: str
    bundle_version: str
    node_counts: dict[str, int] = Field(default_factory=dict)
    edge_counts: dict[str, int] = Field(default_factory=dict)
    routes: dict[str, int] = Field(default_factory=dict)
    gaps: dict[str, Any] = Field(default_factory=dict)
    warnings: list[Warning] = Field(default_factory=list)
    sources: dict[str, Any] = Field(default_factory=dict)
    edges_without_provenance: int = 0
    low_confidence_edges: dict[str, int] = Field(default_factory=dict)
    orphan_nodes: dict[str, Any] = Field(default_factory=dict)
    coverage_summary: dict[str, int] = Field(default_factory=dict)
    ai_candidates: dict[str, Any] = Field(default_factory=dict)
    fatal_errors: list[str] = Field(default_factory=list)


class Bundle(BaseModel):
    bundle_version: str
    generated_at: str
    schema_version: str
    sources: list[Source] = Field(default_factory=list)
    nodes: list[Node] = Field(default_factory=list)
    edges: list[Edge] = Field(default_factory=list)
    indexes: dict[str, Any] = Field(default_factory=dict)
    routes: list[Route] = Field(default_factory=list)
    coverage: Coverage = Field(default_factory=Coverage)
    quality: dict[str, Any] = Field(default_factory=dict)
    provenance: dict[str, Any] = Field(default_factory=dict)
    warnings: list[Warning] = Field(default_factory=list)

File: web/src/test/fixtures/bundle.ts
import type { KnowledgeBundle } from "@/lib/types";

/** Minimal-but-valid KnowledgeBundle fixture for tests (test-only). */
export function makeBundle(): KnowledgeBundle {
  return {
    bundle_version: "test",
    generated_at: "2026-01-01T00:00:00Z",
    schema_version: "test",
    sources: [],
    nodes: [],
    edges: [],
    indexes: {
      by_id: {},
      by_alias: {},
      by_text: {},
      cve_routes: {},
      cwe_to_capec: {},
      capec_to_attack: {},
      attack_to_defend: {},
      attack_to_atlas: {},
      defend_to_controls: {},
      attack_to_detections: {},
      gaps_by_technique: {},
      coverage_by_technique: {},
      sources_by_node: {},
      sources_by_edge: {},
    },
    routes: [],
    coverage: { techniques: [], summary: { covered: 0, partial: 0, gap: 0, unknown: 0, not_applicable: 0 } },
    quality: {
      generated_at: "2026-01-01T00:00:00Z",
      bundle_version: "test",
      node_counts: {},
      edge_counts: {},
      routes: { total: 0, canonical: 0, inferred: 0, framework_total: 0, framework_emitted: 0 },
      gaps: {},
      warnings: [],
      sources: { ok: 0, fallback: 0, unavailable: 0, error: 0, details: [] },
      edges_without_provenance: 0,
      low_confidence_edges: 0,
      orphan_nodes: { total: 0, by_type: {} },
      coverage_summary: { covered: 0, partial: 0, gap: 0, unknown: 0, not_applicable: 0 },
      ai_candidates: {},
      fatal_errors: [],
    },
    provenance: {},
    warnings: [],
  };
}
